src/dust/coagulation/coagulation.hpp

```diff
@@ -0,0 +1,1000 @@
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, mbr, 0, mimax, [&](const int i) {
+    Real sum0 = 0.0;
+    for (int j = 0; j <= i - pgrid - 1; j++) {
+      const Real fett_l = CoagulationRate(i, j, kernel4, vel, stoppingTime, coag, 1);
+      const Real Rf1 = distri(i) * distri(j) * fett_l;
+      const Real dummy = coag.coagR3D(iepsfrag, i, j) * Rf1;
+      sum0 += dummy;
+    }
+
+    if (i - pgrid - 1 >= 0) {
+      Kokkos::atomic_add(&source(i - 1), sum0);
+    }
+
+    Real sum1 = -sum0;
+    // Full fragmentation (only negative terms)
+    int i1 = std::max(0, i - pgrid);
+    for (int j = i1; j <= i; j++) {
+      const Real fett_l = CoagulationRate(i, j, kernel4, vel, stoppingTime, coag, 1);
+      const Real Rf1 = distri(i) * distri(j) * fett_l;
+      sum1 -= Rf1;
+    }
+    Kokkos::atomic_add(&source(i), sum1);
+  });
+  mbr.team_barrier();
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn  void Dust::Coagulation::Coagulation_nQ
+//  \brief
+KOKKOS_FORCEINLINE_FUNCTION
+void Coagulation_nQ(parthenon::team_mbr_t const &mbr, ScratchPad1D<Real> &nQs,
+                    const ScratchPad1D<Real> &Q, const ScratchPad1D<Real> &distri,
+                    const int mimax, const KernelParams &kernel4,
+                    const ScratchPad1D<Real> &vel, const ScratchPad1D<Real> &stoppingTime,
+                    const CoagParams &coag) {
+  // Adding coagulation source terms
+  const int iafrag = coag2drv::afrag;
+  const int iphifrag = coag2drv::phifrag;
+  const int iepsfrag = coag2drv::epsfrag;
+  const int idalp = coag2drv::dalp;
+  const int idpod = coag2drv::dpod;
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, mbr, 0, mimax, [&](const int i) {
+    for (int j = 0; j <= i; j++) {
+      // calculate the rate
+      const Real fett_t = CoagulationRate(i, j, kernel4, vel, stoppingTime, coag, 0);
+      const Real Rc1 = distri(i) * distri(j) * fett_t;
+      const Real &mass_gridi = coag.mass_grid(i);
+      const Real &mass_gridj = coag.mass_grid(j);
+      const Real &dalpji = coag.coagR3D(idalp, j, i);
+      const Real &dpodji = coag.coagR3D(idpod, j, i);
+      const Real &dalpij = coag.coagR3D(idalp, i, j);
+      const Real &dpodij = coag.coagR3D(idpod, i, j);
+
+      const Real Qp1 = Qplus(mass_gridi, Q(i), mass_gridj, Q(j));
+      const Real dphijQ = dalpji * Qp1 * (1.0 + dpodji) - Q(j);
+      const Real dphjiQ = dalpij * Qp1 * (1.0 + dpodij) - Q(i);
+
+      const Real tmp1 = dphijQ - Qp1 * dpodji;
+      const Real tmp2 = dphjiQ - Qp1 * dpodij;
+      for (int nz = 0; nz < 4; nz++) {
+        const int k = coag.cpod_notzero(i, j, nz);
+        if (k < 0) continue;
+        const Real kdeltajk = (j == k) ? 1.0 : 0.0;
+        const Real kdeltaik = (i == k) ? 1.0 : 0.0;
+        const Real nqs_k =
+            (Qp1 * coag.cpod_short(i, j, nz) + tmp1 * kdeltajk + tmp2 * kdeltaik) * Rc1;
+        Kokkos::atomic_add(&nQs(k), nqs_k);
+      }
```

Comment:
```suggestion
      Real nqs_k = 0.0;
      for (int nz = 0; nz < 4; nz++) {
        const int k = coag.cpod_notzero(i, j, nz);
        if (k < 0) continue;
        const Real kdeltajk = (j == k) ? 1.0 : 0.0;
        const Real kdeltaik = (i == k) ? 1.0 : 0.0;
        nqs_k +=
            (Qp1 * coag.cpod_short(i, j, nz) + tmp1 * kdeltajk + tmp2 * kdeltaik) * Rc1;
      }
      Kokkos::atomic_add(&nQs(k), nqs_k);
```
This one should really be a `ScatterView`, but I don't know if you can do that for scratchpads?